moveit_core/robot_model/src/joint_model_group.cpp

```diff
@@ -97,188 +101,239 @@ bool jointPrecedes(const JointModel* a, const JointModel* b)
+    | views::filter(is_mimic_joint)
+    | ranges::to<std::vector>();
+  continuous_joint_model_vector_ = not_fixed_view
+    | views::filter([](const auto& jm) {
+        return jm->getType() == JointModel::REVOLUTE && static_cast<const RevoluteJointModel*>(jm)->isContinuous();
+    })
+    | ranges::to<std::vector>();
+  variable_names_ = not_fixed_view
+    | views::transform(get_variable_names_view)
+    | views::join
+    | ranges::to<std::vector>();
+  variable_names_set_ = variable_names_
+    | ranges::to<std::set>();
+  joint_model_map_ = joint_model_vector_
+    | views::transform([](const auto& jm) { return std::make_pair(jm->getName(), jm); })
+    | ranges::to<std::map>();
+  joint_variables_index_map_ =
+    views::concat(
+      views::zip(not_fixed_view | views::transform(get_variable_names_view),
+                 start_index_view)
+        | views::transform([](const auto& e) {
+          const auto& [names, start_index] = e;
+          return views::zip(names,
+                            views::iota(start_index, start_index + names.size()));
+        })
+        | views::join
+        | views::transform(tuple_to_pair)
+      ,
+      views::zip(not_fixed_view | views::transform(get_name),
+                 start_index_view)
+        | views::transform(tuple_to_pair)
+    )
+    | ranges::to<std::map>();
+  active_joint_models_bounds_ = active_joint_model_vector_
+    | views::transform(get_variable_bounds_addr)
+    | ranges::to<std::vector>();
+  variable_index_list_ =
+    views::zip(not_fixed_view | views::transform(get_first_variable_index),
+               not_fixed_view | views::transform(get_variable_names_size))
+    | views::transform([](const auto& e) {
+     const auto& [first_index, names_size] = e;
+     return views::iota(first_index, first_index + names_size);
+    })
+    | views::join
+    | ranges::to<std::vector>();
+  active_joint_model_start_index_ =
+    views::zip(start_index_view, not_fixed_view)
+    | views::remove_if([](const auto& e) { return is_mimic_joint(std::get<1>(e)); })
+    | views::transform([](const auto& e) { return std::get<0>(e); })
+    | ranges::to<std::vector>();
+  variable_count_ = ranges::accumulate(not_fixed_view | views::transform(get_variable_count), 0U);
+  active_variable_count_ = ranges::accumulate(not_fixed_view | views::transform(get_variable_count), 0U);
+  is_single_dof_ = ranges::none_of(not_fixed_view | views::transform(get_variable_count),
+    [](const auto& e) { return e > 1; });
 
   // now we need to find all the set of joints within this group
   // that root distinct subtrees
-  for (const JointModel* active_joint_model : active_joint_model_vector_)
-  {
-    // if we find that an ancestor is also in the group, then the joint is not a root
-    if (!includesParent(active_joint_model, this))
-      joint_roots_.push_back(active_joint_model);
-  }
+  joint_roots_ = active_joint_model_vector_
+    | views::filter([this](const auto& jm) {
```

Comment:
Do you think you should consider `remove_if` to avoid the negation of `includesParent` or does that result in something different?

Reply:
good catch, changed